patch/tez/tez-0.9.1_configurable_edge_info_support.patch

```diff
@@ -0,0 +1,532 @@
+ import org.apache.tez.runtime.library.common.ConfigUtils;
++import org.apache.tez.runtime.library.common.RssShuffleFactory;
+ import org.apache.tez.runtime.library.output.UnorderedKVOutput;
+
+ @InterfaceAudience.Public
+@@ -222,9 +223,17 @@ public Builder setAdditionalConfiguration(Map<String, String> confMap) {
+     public Builder setFromConfiguration(Configuration conf) {
+       // Maybe ensure this is the first call ? Otherwise this can end up overriding other parameters
+       Preconditions.checkArgument(conf != null, "Configuration cannot be null");
+-      Map<String, String> map = ConfigUtils.extractConfigurationMap(conf,
+-          Lists.newArrayList(UnorderedKVOutput.getConfigurationKeySet(),
+-              TezRuntimeConfiguration.getRuntimeAdditionalConfigKeySet()), TezRuntimeConfiguration.getAllowedPrefixes());
++      Map<String, String> map;
++      if (RssShuffleFactory.isRssEnabled(conf)) {
++        map = ConfigUtils.extractConfigurationMap(conf,
++                Lists.newArrayList(RssShuffleFactory.RssUnorderedKVOutput.getConfigurationKeySet(),
++                        TezRuntimeConfiguration.getRuntimeAdditionalConfigKeySet()), TezRuntimeConfiguration.getAllowedPrefixes());
++      } else {
++        map = ConfigUtils.extractConfigurationMap(conf,
++                Lists.newArrayList(UnorderedKVOutput.getConfigurationKeySet(),
++                        TezRuntimeConfiguration.getRuntimeAdditionalConfigKeySet()), TezRuntimeConfiguration.getAllowedPrefixes());
++      }
++
+       ConfigUtils.addConfigMapToConfiguration(this.conf, map);
+       return this;
+     }
+diff --git a/tez-runtime-library/src/main/java/org/apache/tez/runtime/library/conf/UnorderedPartitionedKVEdgeConfig.java b/tez-runtime-library/src/main/java/org/apache/tez/runtime/library/conf/UnorderedPartitionedKVEdgeConfig.java
+index 52da491517..d7cb581023 100644
+--- a/tez-runtime-library/src/main/java/org/apache/tez/runtime/library/conf/UnorderedPartitionedKVEdgeConfig.java
++++ b/tez-runtime-library/src/main/java/org/apache/tez/runtime/library/conf/UnorderedPartitionedKVEdgeConfig.java
+@@ -34,6 +34,7 @@
+ import org.apache.tez.dag.api.InputDescriptor;
+ import org.apache.tez.dag.api.OutputDescriptor;
+ import org.apache.tez.dag.api.UserPayload;
++import org.apache.tez.runtime.library.common.RssShuffleFactory;
+ import org.apache.tez.runtime.library.input.UnorderedKVInput;
+ import org.apache.tez.runtime.library.output.UnorderedPartitionedKVOutput;
+
+@@ -99,6 +100,9 @@ public UserPayload getOutputPayload() {
+
+   @Override
+   public String getOutputClassName() {
++    if(RssShuffleFactory.isRssEnabled(getOutputPayload())){
++      return RssShuffleFactory.RssUnOrderedPartitionedKVOutput.outputClassName;
++    }
+     return UnorderedPartitionedKVOutput.class.getName();
+   }
+
+@@ -119,6 +123,9 @@ public String getInputHistoryText() {
+
+   @Override
+   public String getInputClassName() {
++    if(RssShuffleFactory.isRssEnabled(getInputPayload())){
```

Comment:
Could we get the class name from the configuration? 
We could add a new config option like `tez.input.class.name`.
We return conf.get("tez.input.class.name"). The default value of this config option is the value of `UnorderedKVInput.class.getName()`.
We hope we can remove `RssShuffleFactory`. We hope reduce the size of this patch.  We hope this patch become more common. This should be a plugin mechism like MapReduce. It will be used for custom shuffle service. It shouldn't only bind remote shuffle service.

Reply:
Ok, I'll see how to modify it these days